File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/geometry/GeometryEngineController.java
package com.valentingrigorean.arcgis_maps_flutter.geometry;

import android.util.Log;

import androidx.annotation.NonNull;

import com.esri.arcgisruntime.geometry.AngularUnit;
import com.esri.arcgisruntime.geometry.AngularUnitId;
import com.esri.arcgisruntime.geometry.GeodesicSectorParameters;
import com.esri.arcgisruntime.geometry.GeodeticCurveType;
import com.esri.arcgisruntime.geometry.GeodeticDistanceResult;
import com.esri.arcgisruntime.geometry.Geometry;
import com.esri.arcgisruntime.geometry.GeometryEngine;
import com.esri.arcgisruntime.geometry.LinearUnit;
import com.esri.arcgisruntime.geometry.LinearUnitId;
import com.esri.arcgisruntime.geometry.Point;
import com.esri.arcgisruntime.geometry.Polygon;
import com.esri.arcgisruntime.geometry.SpatialReference;
import com.valentingrigorean.arcgis_maps_flutter.Convert;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class GeometryEngineController implements MethodChannel.MethodCallHandler {
    private static final String TAG = "GeometryEngineController";
    private final MethodChannel channel;

    public GeometryEngineController(BinaryMessenger messenger) {
        channel = new MethodChannel(messenger, "plugins.flutter.io/arcgis_channel/geometry_engine");
        channel.setMethodCallHandler(this);
    }

    public void dispose() {
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "project": {
                handleProject(call.arguments(), result);
            }
            break;
            case "distanceGeodetic": {
                final Map<?, ?> data = Convert.toMap(call.arguments);
                handleDistanceGeodetic(data, result);
            }
            break;
            case "bufferGeometry": {
                handleBufferGeometry(call.arguments(), result);
            }
            break;
            case "geodeticBufferGeometry": {
                handleGeodeticBufferGeometry(call.arguments(), result);
            }
            break;
            case "intersection": {
                handleIntersection(call.arguments(), result);
            }
            break;
            case "intersections": {
                handleIntersections(call.arguments(), result);
            }
            break;
            case "geodesicSector": {
                final GeodesicSectorParameters params = Convert.toGeodesicSectorParameters(call.arguments);
                final Geometry geometry = GeometryEngine.sectorGeodesic(params);
                result.success(Convert.geometryToJson(geometry));
            }
            break;
            case "geodeticMove": {
                handleGeodeticMove(call.arguments(), result);
            }
            case "simplify": {
                handleSimply(call.arguments(), result);
            }
            case "isSimple": {
                handleIsSimple(call.arguments(), result);
            }
            break;
            default:
                result.notImplemented();
                break;
        }
    }

    private void handleGeodeticMove(Map<?, ?> data, MethodChannel.Result result) {
        final List<?> rawPoints = Convert.toList(data.get("points"));
        final double distance = Convert.toDouble(data.get("distance"));
        final LinearUnitId distanceUnitId = Convert.toLinearUnitId(data.get("distanceUnit"));
        final double azimuth = Convert.toDouble(data.get("azimuth"));
        final AngularUnitId azimuthUnitId = Convert.toAngularUnitId(data.get("azimuthUnit"));
        final GeodeticCurveType curveType = Convert.toGeodeticCurveType(data.get("curveType"));

        final List<Point> points = new ArrayList<>();
        for (Object rawPoint : rawPoints) {
            points.add(Convert.toPoint(rawPoint));
        }

        final List<Point> movedPoints = GeometryEngine.moveGeodetic(points, distance, new LinearUnit(distanceUnitId), azimuth, new AngularUnit(azimuthUnitId), curveType);
        final List<Object> resultPoints = new ArrayList<>();
        for (Point point : movedPoints) {
            resultPoints.add(Convert.geometryToJson(point));
        }
        result.success(resultPoints);
    }

    private void handleIntersections(Map<?, ?> data, MethodChannel.Result result) {

        final Geometry firstGeometry = Convert.toGeometry(data.get("firstGeometry"));
        final Geometry secondGeometry = Convert.toGeometry(data.get("secondGeometry"));
        List<Geometry> geometryList = GeometryEngine.intersections(firstGeometry, secondGeometry);
        ArrayList<Object> resultList = new ArrayList<>();
        for (Geometry geometry : geometryList) {
            resultList.add(Convert.geometryToJson(geometry));
        }
        result.success(resultList);
    }

    private void handleIntersection(Map<?, ?> data, MethodChannel.Result result) {
        final Geometry firstGeometry = Convert.toGeometry(data.get("firstGeometry"));
        final Geometry secondGeometry = Convert.toGeometry(data.get("secondGeometry"));
        Geometry geometry = GeometryEngine.intersection(firstGeometry, secondGeometry);
        result.success(Convert.geometryToJson(geometry));
    }

    private void handleGeodeticBufferGeometry(Map<?, ?> data, MethodChannel.Result result) {
        final Geometry geometry = Convert.toGeometry(data.get("geometry"));
        final double distance = Convert.toDouble(data.get("distance"));
        final LinearUnitId distanceUnitId = Convert.toLinearUnitId(data.get("distanceUnit"));
        final double maxDeviation = Convert.toDouble(data.get("maxDeviation"));
        final GeodeticCurveType curveType = Convert.toGeodeticCurveType(data.get("curveType"));
        final Polygon polygon = GeometryEngine.bufferGeodetic(geometry, distance, new LinearUnit(distanceUnitId), maxDeviation, curveType);
        if (polygon == null) {
            result.success(null);
        } else {
            result.success(Convert.geometryToJson(polygon));
        }
    }

    private void handleBufferGeometry(Map<?, ?> data, MethodChannel.Result result) {
        final Geometry geometry = Convert.toGeometry(data.get("geometry"));
        final double distance = Convert.toDouble(data.get("distance"));
        final Polygon polygon = GeometryEngine.buffer(geometry, distance);
        if (polygon == null) {
            result.success(null);
        } else {
            result.success(Convert.geometryToJson(polygon));
        }
    }

    private void handleProject(Map<?, ?> data, MethodChannel.Result result) {
        final SpatialReference spatialReference = Convert.toSpatialReference(data.get("spatialReference"));
        final Geometry geometry = Convert.toGeometry(data.get("geometry"));
        final Geometry projectedGeometry = GeometryEngine.project(geometry, spatialReference);
        if (projectedGeometry == null) {
            result.success(null);
        } else {
            result.success(Convert.geometryToJson(projectedGeometry));
        }
    }

    private void handleDistanceGeodetic(Map<?, ?> data, MethodChannel.Result result) {
        final Point point1 = (Point) Convert.toGeometry(data.get("point1"));
        final Point point2 = (Point) Convert.toGeometry(data.get("point2"));
        final LinearUnitId distanceUnitId = Convert.toLinearUnitId(data.get("distanceUnitId"));
        final AngularUnitId azimuthUnitId = Convert.toAngularUnitId(data.get("azimuthUnitId"));
        final GeodeticCurveType curveType = Convert.toGeodeticCurveType(data.get("curveType"));
        try {
            final GeodeticDistanceResult geodeticDistanceResult = GeometryEngine.distanceGeodetic(point1, point2, new LinearUnit(distanceUnitId), new AngularUnit(azimuthUnitId), curveType);
            if (geodeticDistanceResult == null) {
                result.success(null);
                return;
            }
            final Map<String, Object> json = new HashMap<>(4);
            json.put("distance", geodeticDistanceResult.getDistance());
            json.put("distanceUnitId", data.get("distanceUnitId"));
            json.put("azimuth1", geodeticDistanceResult.getAzimuth1());
            json.put("azimuth2", geodeticDistanceResult.getAzimuth2());
            json.put("angularUnitId", data.get("azimuthUnitId"));
            result.success(json);
        } catch (Exception ex) {
            Log.e(TAG, "Failed to get distanceGeodetic ", ex);
            result.success(null);
        }
    }

    private void handleSimply(Map<?, ?> data, MethodChannel.Result result) {
        Geometry originGeometry = Convert.toGeometry(data);
        if (originGeometry == null) {
            Log.e(TAG, "Failed to simply as geometry is null");
            result.success(null);
        } else {
            Geometry simplifiedGeometry = GeometryEngine.simplify(originGeometry);
            result.success(Convert.geometryToJson(simplifiedGeometry));
        }
    }

    private void handleIsSimple(Map<?, ?> data, MethodChannel.Result result) {
        Geometry originGeometry = Convert.toGeometry(data);
        if (originGeometry == null) {
            Log.e(TAG, "Failed to get isSimple as geometry is null");
            result.success(true);
        } else {
            result.success(GeometryEngine.isSimple(originGeometry));
        }
    }
}

File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/geometry/GeometryEngineController.kt
package com.valentingrigorean.arcgis_maps_flutter.geometry

import android.util.Log
import com.arcgismaps.geometry.AngularUnit
import com.arcgismaps.geometry.AreaUnit
import com.arcgismaps.geometry.GeometryEngine
import com.arcgismaps.geometry.LinearUnit
import com.arcgismaps.geometry.Point
import com.valentingrigorean.arcgis_maps_flutter.convert.geometry.*
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

class GeometryEngineController(messenger: BinaryMessenger) : MethodCallHandler {
    private val channel: MethodChannel

    init {
        channel = MethodChannel(messenger, "plugins.flutter.io/arcgis_channel/geometry_engine")
        channel.setMethodCallHandler(this)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "project" -> {
                handleProject(call.arguments()!!, result)
            }

            "distanceGeodetic" -> {
                handleDistanceGeodetic(call.arguments()!!, result)
            }

            "bufferGeometry" -> {
                handleBufferGeometry(call.arguments()!!, result)
            }

            "geodeticBufferGeometry" -> {
                handleGeodeticBufferGeometry(call.arguments()!!, result)
            }

            "intersection" -> {
                handleIntersection(call.arguments()!!, result)
            }

            "intersections" -> {
                handleIntersections(call.arguments()!!, result)
            }

            "geodesicSector" -> {
                val params = call.arguments.toGeodesicSectorParametersOrNull()!!
                val geometry = GeometryEngine.sectorGeodesicOrNull(params)
                result.success(geometry?.toFlutterJson())
            }

            "geodeticMove" -> {
                handleGeodeticMove(call.arguments()!!, result)
            }

            "simplify" -> {
                handleSimply(call.arguments()!!, result)
            }

            "isSimple" -> {
                handleIsSimple(call.arguments()!!, result)
            }

            "densifyGeodetic" -> {
                handleDensifyGeodetic(call.arguments()!!, result)
            }

            "lengthGeodetic" -> {
                handleLengthGeodetic(call.arguments()!!, result)
            }

            "areaGeodetic" -> {
                handleAreaGeodetic(call.arguments()!!, result)
            }

            "getExtent" -> {
                handleGetExtent(call.arguments()!!, result)
            }

            else -> result.notImplemented()
        }
    }

    private fun handleProject(data: Map<*, *>, result: MethodChannel.Result) {
        val spatialReference = data["spatialReference"]?.toSpatialReferenceOrNull()!!
        val geometry = data["geometry"]?.toGeometryOrNull()!!
        val projectedGeometry = GeometryEngine.projectOrNull(geometry, spatialReference)
        result.success(projectedGeometry?.toFlutterJson())
    }

    private fun handleDistanceGeodetic(data: Map<*, *>, result: MethodChannel.Result) {
        val point1 = data["point1"]?.toGeometryOrNull() as Point
        val point2 = data["point2"]?.toGeometryOrNull() as Point
        val distanceUnitId = (data["distanceUnitId"] as Int).toLinearUnitId()
        val azimuthUnitId = (data["azimuthUnitId"] as Int).toAngularUnitId()
        val curveType = (data["curveType"] as Int).toGeodeticCurveType()
        try {
            val geodeticDistanceResult = GeometryEngine.distanceGeodeticOrNull(
                point1,
                point2,
                LinearUnit(distanceUnitId),
                AngularUnit(azimuthUnitId),
                curveType
            )
            result.success(geodeticDistanceResult?.toFlutterJson())
        } catch (ex: Exception) {
            Log.e(TAG, "Failed to get distanceGeodetic ", ex)
            result.success(null)
        }
    }

    private fun handleBufferGeometry(data: Map<*, *>, result: MethodChannel.Result) {
        val geometry = data["geometry"]?.toGeometryOrNull()!!
        val distance = data["distance"] as Double
        val polygon = GeometryEngine.bufferOrNull(geometry, distance)
        result.success(polygon?.toFlutterJson())
    }

    private fun handleGeodeticBufferGeometry(data: Map<*, *>, result: MethodChannel.Result) {
        val geometry = data["geometry"]?.toGeometryOrNull()!!
        val distance = data["distance"] as Double
        val distanceUnitId = (data["distanceUnit"] as Int).toLinearUnitId()
        val maxDeviation = data["maxDeviation"] as Double
        val curveType = (data["curveType"] as Int).toGeodeticCurveType()
        val polygon = GeometryEngine.bufferGeodeticOrNull(
            geometry,
            distance,
            LinearUnit(distanceUnitId),
            maxDeviation,
            curveType
        )
        result.success(polygon?.toFlutterJson())
    }

    private fun handleIntersection(data: Map<*, *>, result: MethodChannel.Result) {
        val firstGeometry = data["firstGeometry"]?.toGeometryOrNull()!!
        val secondGeometry = data["secondGeometry"]?.toGeometryOrNull()!!
        val geometry = GeometryEngine.intersectionOrNull(firstGeometry, secondGeometry)
        result.success(geometry?.toFlutterJson())
    }

    private fun handleGeodeticMove(data: Map<*, *>, result: MethodChannel.Result) {
        val points = (data["points"] as List<*>).map { it?.toPointOrNull()!! }
        val distance = data["distance"] as Double
        val distanceUnitId = (data["distanceUnit"] as Int).toLinearUnitId()
        val azimuth = data["azimuth"] as Double
        val azimuthUnitId = (data["azimuthUnit"] as Int).toAngularUnitId()
        val curveType = (data["curveType"] as Int).toGeodeticCurveType()
        val movedPoints = GeometryEngine.tryMoveGeodetic(
            points,
            distance,
            LinearUnit(distanceUnitId),
            azimuth,
            AngularUnit(azimuthUnitId),
            curveType
        )
        result.success(movedPoints.map { it.toFlutterJson() })
    }

    private fun handleSimply(data: Map<*, *>, result: MethodChannel.Result) {
        val originGeometry = data?.toGeometryOrNull()!!
        val simplifiedGeometry = GeometryEngine.simplifyOrNull(originGeometry)
        result.success(simplifiedGeometry?.toFlutterJson())
    }

    private fun handleIntersections(data: Map<*, *>, result: MethodChannel.Result) {
        val firstGeometry = data["firstGeometry"]?.toGeometryOrNull()!!
        val secondGeometry = data["secondGeometry"]?.toGeometryOrNull()!!
        val geometryList = GeometryEngine.tryIntersections(firstGeometry, secondGeometry)
        result.success(geometryList.map { it.toFlutterJson() })
    }


    private fun handleIsSimple(data: Map<*, *>, result: MethodChannel.Result) {
        val originGeometry = data?.toGeometryOrNull()!!
        result.success(GeometryEngine.isSimple(originGeometry))
    }

    private fun handleDensifyGeodetic(data: Map<*, *>, result: MethodChannel.Result) {
        val geometry = data["geometry"]?.toGeometryOrNull()!!
        val maxSegmentLength = data["maxSegmentLength"] as Double
        val linearUnitId = (data["lengthUnit"] as Int).toLinearUnitId()
        val curveType = (data["curveType"] as Int).toGeodeticCurveType()
        val resultGeometry = GeometryEngine.densifyGeodeticOrNull(
            geometry,
            maxSegmentLength,
            LinearUnit(linearUnitId),
            curveType
        )
        result.success(resultGeometry?.toFlutterJson())
    }

    private fun handleLengthGeodetic(data: Map<*, *>, result: MethodChannel.Result) {
        val geometry = data["geometry"]?.toGeometryOrNull()!!
        val linearUnitId = (data["lengthUnit"] as Int).toLinearUnitId()
        val curveType = (data["curveType"] as Int).toGeodeticCurveType()
        result.success(
            GeometryEngine.lengthGeodetic(
                geometry,
                LinearUnit(linearUnitId),
                curveType
            )
        )
    }

    private fun handleAreaGeodetic(data: Map<*, *>, result: MethodChannel.Result) {
        val geometry = data["geometry"]?.toGeometryOrNull()!!
        val areaUnit = (data["areaUnit"] as Int).toAreaUnitId()
        val curveType = (data["curveType"] as Int).toGeodeticCurveType()
        result.success(GeometryEngine.areaGeodetic(geometry, AreaUnit(areaUnit), curveType))
    }

    private fun handleGetExtent(data: Map<*, *>, result: MethodChannel.Result) {
        val geometry = data["geometry"]?.toGeometryOrNull()!!
        result.success(geometry.extent.toFlutterJson())
    }

    companion object {
        private const val TAG = "GeometryEngineController"
    }
}